### PR TITLE
E2E design and implementation: Slack integration

### PR DESCRIPTION
**Pull request by mauromalara**
|Related issue|
|-------------|
| #3126 |

## Description

This Pull Request adds a test that checks if Wazuh sends the alert via Slack by attempting a brute force SSH attack.

The only precondition is to create an inventory as follows:
```
all:
  hosts:
    wazuh-manager:
      ansible_connection: ssh
      ansible_user: <VM USER>
      ansible_ssh_private_key_file: <PRIVATE KEY PATH>
      ansible_python_interpreter: <PYTHON INTERPRETER PATH>
      dashboard_user: <USER>
      dashboard_password: <PASS>
  vars:
    web_hook_url: <SLACK_WEBHOOK_URL>
    slack_token: <TOKEN>
    slack_channel: <CHANNEL_ID>
```

## Use cases

- Test that Wazuh sends the alert via Slack by attempting a brute force SSH attack (alert of level 10).

### Added

- `test_slack_integration.py`: this module test the integration with Slack.

---

## Testing performed [Code analysis](https://github.com/wazuh/wazuh-qa/runs/7567775946?check_suite_focus=true)

| Tester             | Test path | Jenkins | Local  | OS | Commit | Notes                |
|--------------------|-----------|---------|--------|-----|--------|----------------------|
| @mauromalara (Developer)  |     test_suricata_integration/      |  :no_entry_sign: :no_entry_sign: :no_entry_sign: | [🟢](https://github.com/wazuh/wazuh-qa/files/9303018/r1-CentOS-manager-3126.zip) [🟢](https://github.com/wazuh/wazuh-qa/files/9303019/r2-CentOS-manager-3126.zip) [🟢](https://github.com/wazuh/wazuh-qa/files/9303020/r3-CentOS-manager-3126.zip) | CentOS | 05475bc | Nothing to highlight |
| @juliamagan (Reviewer)   |     test_suricata_integration/      |  :no_entry_sign: :no_entry_sign: :no_entry_sign:| [🟢 ](https://github.com/wazuh/wazuh-qa/files/9268897/R1-3136-test-slack-integration.zip)[🟢 ](https://github.com/wazuh/wazuh-qa/files/9268904/R2-3136-test-slack-integration.zip)[🟢 ](https://github.com/wazuh/wazuh-qa/files/9268915/R3-3136-test-slack-integration.zip)  |  CentOS      |    a6d3836    | Nothing to highlight |

### CODE REVIEW

File: tests/end_to_end/test_slack_integration/data/playbooks/configuration.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    - name: Truncate alerts log
+      shell: echo "" > {{ alerts_path }}
+
+    - name: Restart wazuh-manager
+      systemd:
+        state: restarted
+        name: wazuh-manager
+
+    - name: Install pexpect using pip
+      shell: python3 -m pip install pexpect>=3.3
```
Comment (juliamagan) [inline tests/end_to_end/test_slack_integration/data/playbooks/configuration.yaml:31]: Why is `pexpect` needed?

File: tests/end_to_end/test_slack_integration/data/playbooks/generate_events.yaml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  tasks:
+
+    - name: Truncate integrations log
+      shell: echo "" > {{ integrations_log }}
+
+    - name: Attempt a brute force SSH attack
+      expect:
+        command: ssh {{ item }}@localhost
+        timeout: 5
+        responses:
```
Comment (juliamagan) [inline tests/end_to_end/test_slack_integration/data/playbooks/generate_events.yaml:16]: We must take into consideration clean environments and the fingerprint.

File: tests/end_to_end/test_slack_integration/data/playbooks/generate_events.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      loop:
+        - not-a-user
+        - not-a-user
+        - not-a-user
+        - not-a-user
+        - not-a-user
+        - not-a-user
+        - not-a-user
+        - not-a-user
+      ignore_errors: true
```
Comment (juliamagan) [inline tests/end_to_end/test_slack_integration/data/playbooks/generate_events.yaml:27]: We must ignore specific errors, not all of them.

**Review by juliamagan — CHANGES_REQUESTED** [review]

File: tests/end_to_end/test_slack_integration/data/test_cases/cases_slack_integration.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+- name: Slack integration - SSH Brute force attack
```
Comment (juliamagan) [inline tests/end_to_end/test_slack_integration/data/test_cases/cases_slack_integration.yaml:1]: Spaces in the test case name are not recommended.

**Review by juliamagan — CHANGES_REQUESTED** [review]

File: tests/end_to_end/test_slack_integration/data/test_cases/cases_slack_integration.yaml (hunk 4 above)
Comment (mauromalara) [reply tests/end_to_end/test_slack_integration/data/test_cases/cases_slack_integration.yaml]: Done in 6c4f812

File: tests/end_to_end/test_slack_integration/data/playbooks/generate_events.yaml (hunk 3 above)
Comment (mauromalara) [reply tests/end_to_end/test_slack_integration/data/playbooks/generate_events.yaml]: Done in 6c4f812

File: tests/end_to_end/test_slack_integration/data/playbooks/generate_events.yaml (hunk 2 above)
Comment (mauromalara) [reply tests/end_to_end/test_slack_integration/data/playbooks/generate_events.yaml:16]: Done in 6c4f812

File: tests/end_to_end/test_slack_integration/data/playbooks/configuration.yaml (hunk 1 above)
Comment (mauromalara) [reply tests/end_to_end/test_slack_integration/data/playbooks/configuration.yaml]: Done in 6c4f812

**Review by juliamagan — CHANGES_REQUESTED** [review]
We need to check the alert through Slack API

File: tests/end_to_end/test_slack_integration/test_slack_integration.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+# Custom paths
+slack_api_script = os.path.join(test_data_path, 'configuration', 'slack_api_script.py')
+
+# Update configuration with custom paths
+metadata = config.update_configuration_template(metadata, ['CUSTOM_SLACK_SCRIPT_PATH'], [slack_api_script])
+
+
+@pytest.fixture(scope='function')
+def get_slack_log_path():
```
Comment (juliamagan) [inline tests/end_to_end/test_slack_integration/test_slack_integration.py:35]: The path should be defined above and the fixture should only delete the file.

**Review by juliamagan — CHANGES_REQUESTED** [review]

File: tests/end_to_end/test_slack_integration/test_slack_integration.py (hunk 5 above)
Comment (mauromalara) [reply tests/end_to_end/test_slack_integration/test_slack_integration.py]: Done in 29a7187

**Review by juliamagan — APPROVED** [review]

File: tests/end_to_end/test_slack_integration/data/configuration/slack_api_script.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+def get_parameters():
+    """
+    Returns:
+        argparse.Namespace: Object with the user parameters.
+    """
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument('--token', '-t', type=str, action='store', required=True)
+    parser.add_argument('--channel', '-c', type=str, action='store', required=True)
+    parser.add_argument('--messages_limit', '-m', type=int, action='store', required=True)
```
Comment (jmv74211) [inline tests/end_to_end/test_slack_integration/data/configuration/slack_api_script.py:17]:
```suggestion
    parser.add_argument('--messages-limit', '-m', type=int, action='store', required=True)
```

File: tests/end_to_end/test_slack_integration/data/configuration/slack_api_script.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    parser.add_argument('--path', '-p', type=str, action='store', required=True)
+
+    arguments = parser.parse_args()
+
+    return arguments
+
+
+def main():
+    parameters = get_parameters()
+
```
Comment (jmv74211) [inline tests/end_to_end/test_slack_integration/data/configuration/slack_api_script.py:27]: Add comments during the different sections

**Review by jmv74211 — CHANGES_REQUESTED** [review]

File: tests/end_to_end/test_slack_integration/data/configuration/slack_api_script.py (hunk 7 above)
Comment (mauromalara) [reply tests/end_to_end/test_slack_integration/data/configuration/slack_api_script.py:28]: Done in 05475bc

File: tests/end_to_end/test_slack_integration/data/configuration/slack_api_script.py (hunk 6 above)
Comment (mauromalara) [reply tests/end_to_end/test_slack_integration/data/configuration/slack_api_script.py]: Done in 05475bc

**Review by jmv74211 — APPROVED** [review]

**Pull request merged**